### PR TITLE
[BUGFIX] Ensure Iterables work with dictionaries

### PR DESCRIPTION
**Pull request by pzuraq**
In the recent VM upgrade, a general API for adding debug information to
references was added. This included debug information for iteration
items, which have dynamic path segments. There were two issues with the
implementation:

1. We were using the `key` instead of the `memo` property for displaying
   which item in the collection was being used. Counterintuitively,
   `memo` is the actual key for the value in the collection, whereas
   `key` is the memoization key used to determine if the item should be
   updated. This has been updated, so that `memo` is used to show users
   where they are in the collection.

2. Since keys can be arbitrary objects, we can't always safely coerce
   them to be strings, so we have to add some debug-only logic for
   coercion.

### CODE REVIEW

**Comment by mansona** [comment]
it looks like this is passing and it's just linting that is causing CI to fail 🤔 does that mean that the bug is fixed!? 🎉

File: packages/@glimmer/util/lib/debug-to-string.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      string = `(an instance of ${className})`;
+    } else if (value === undefined) {
+      string = '(an unknown tag)';
+    } else {
+      string = String(value);
+    }
+
+    return string;
+  };
+}
```
Comment (chancancode) [inline packages/@glimmer/util/lib/debug-to-string.ts]: Should this just be a no-op function in prod? The module can't be removed anyway so I don't know if it's worth making every caller care about it. If it matters (probably not), we should just make it a macro.

File: packages/@glimmer/util/lib/debug-to-string.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    let string;
+
+    if (typeof value === 'function') {
+      string = value.name;
+    } else if (typeof value === 'object' && value !== null) {
+      let className = (value.constructor && value.constructor.name) || '(unknown class)';
+
+      string = `(an instance of ${className})`;
+    } else if (value === undefined) {
+      string = '(an unknown tag)';
```
Comment (chancancode) [inline packages/@glimmer/util/lib/debug-to-string.ts]: Why isn't this `"undefined"`?

File: packages/@glimmer/util/lib/debug-to-string.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    } else {
+      string = String(value);
```
Comment (chancancode) [inline packages/@glimmer/util/lib/debug-to-string.ts]: try/catch?

**Comment by pzuraq** [comment]
Yup, exactly. I've updated the function to be the same as Ember's implementation.

**Pull request merged**